crates/leon/src/error.rs

```diff
@@ -0,0 +1,131 @@
+#[cfg(feature = "miette")]
+use miette::{Diagnostic, SourceSpan};
+use thiserror::Error;
+
+#[cfg_attr(feature = "miette", derive(Diagnostic))]
+#[derive(Debug, Error)]
+pub enum RenderError {
+    /// A key was missing from the provided values.
+    #[error("missing key `{0}`")]
+    MissingKey(String),
+
+    /// An I/O error passed through from [`Template::render_into`].
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+/// An error that can occur when parsing a template.
+///
+/// This is a rich miette-powered error which will highlight the source of the
+/// error in the template when output (with miette's `fancy` feature). If you
+/// don't want or need that, you can disable the `miette` feature and a simpler
+/// opaque error will be substituted.
+#[cfg(feature = "miette")]
+#[derive(Clone, Debug, Error, PartialEq, Eq)]
+#[error(transparent)]
+pub struct ParseError(pub Box<InnerParseError>);
+
+#[cfg(feature = "miette")]
+impl Diagnostic for ParseError {
+    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
+        self.0.source_code()
+    }
+
+    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
+        self.0.labels()
+    }
+}
+
+/// The inner (unboxed) type of [`ParseError`].
+#[cfg(feature = "miette")]
+#[derive(Clone, Debug, Diagnostic, Error, PartialEq, Eq)]
+#[error("template parse failed")]
+#[non_exhaustive]
+pub struct InnerParseError {
```

Comment:
I don't think it makes sense to expose `InnerParseError` as public:

```suggestion
struct InnerParseError {
```